flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/TwoPhaseCommittingStatefulSink.java

```diff
@@ -0,0 +1,34 @@
+package org.apache.flink.connector.kafka.sink;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.connector.sink2.StatefulSink;
+import org.apache.flink.api.connector.sink2.TwoPhaseCommittingSink;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * A combination of {@link TwoPhaseCommittingSink} and {@link StatefulSink}.
+ *
+ * <p>The purpose of this interface is to be able to pass an interface rather than a {@link
+ * KafkaSink} implementation into the reducing sink which simplifies unit testing.
```

Comment:
I guess it's probably not only for unit testing? Without the mixin interface the `ReducingUpsertSink` and `ReducingUpsertWriter` wouldn't be able to properly wrap the KafkaSink / KafkaWriter delegates.